geth/common/queued_tx.go

```diff
@@ -0,0 +1,144 @@
+package common
+
+import (
+	"context"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// QueuedTx holds enough information to complete the queued transaction.
+type QueuedTx struct {
```

Comment:
I am not really sure it's a good idea to design this struct this way. It does not scale. If you had ten different structs, you would need to create 10*N getters and setters. It was better to leave a mutex in `TxQueue`.

Reply:
It is not safe to make thread unsafe methods and give to client responsobility to hold safaty.

Reply:
Why not? This is quite common as a majority of structs in Go standard library is not thread-safe. Thinking this way each struct should have only private fields and getters/setters wrapped with a mutex.

Reply:
This code had has 5 data races in itself and 3 data races in tests. I think it is strong argument to make `QueuedTx` thread safe - no one wants data race in transaction.

Reply:
Hmm, so am I right that the concurrent access to the information of the former struct led to the races and now the struct semantically still is only a DTO but also ensures synchronisation about the mutex?

As it is on the accessor level, we have no situations _set A; get A; set B_ vs _set A; set B; get A_?

Reply:
I think it little bit moree than DTO, it has state inProgress that can be changed in different goroutines, done and discard channels for logic synchronisation.
Both setuations can get place.